### PR TITLE
Add more functions for ensuring nodes via MastForestBuilder

### LINKED ISSUE #1355: `MastForest`: Add convenience method for tests
_Originally posted by @bobbinth in https://github.com/0xPolygonMiden/miden-vm/pull/1349#discussion_r1640372101_          
          
Not for this PR, but I see this pattern a lot in tests. I wonder if it makes sense to create convenience methods on `MastForest` which would create a node and add it in a single call. For example, something like:
```Rust
let last_basic_block_id = mast_forest.ensure_basic_block(vec![Operation::FmpAdd]);
```

### PR DESCRIPTION
**Pull request by sergerad**
Closes #1355

Add methods to`MastForestBuilder` to facilitate ensuring nodes with fewer LOC

### CODE REVIEW

**Comment by sergerad** [comment]
Hey @plafer reviewers aren't assigned automatically and I can't request reviews. This one is ready for review.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you for tackling this one as well!

Overall, I think the solution works but we probably want to add convenience methods for other nodes too (e.g., for `Join`, and `Split` nodes). For these, I don't think we can use the `Into` approach (or at least I'm not sure how). So, we may end up needed `MastForestBuilder::ensure_join()`, `MastForestBuilder::ensure_split()` etc.

If we do the above, we might want to have a similar approach for basic blocks. Though, I agree, `MastForestBuilder::ensure_basic_block()` does not look as elegant as the current solution. Maybe, we can just say `MastForestBuilder::ensure_block()`.

Separately, and not for this PR, I find the fact that we need to say "basic_block" in various places pretty cumbersome. Maybe we should just say "block" everywhere instead. @plafer and @bitwalker, curious what you think on this.

**Comment by bobbinth** [comment]
Here is an [example](https://github.com/0xPolygonMiden/miden-vm/blob/88eb223e09297147eee74c08cbf4b3998c0ae76f/assembly/src/assembler/tests.rs#L196-L199) where instead of:
```rust
let nested = {
    let node = MastNode::new_split(r#true2, r#false2, expected_mast_forest_builder.forest());
    expected_mast_forest_builder.ensure_node(node)
};
```
We should be able to do something like:
```rust
let nested = expected_mast_forest_builder.ensure_split(r#true2, r#false2);
```

**Comment by sergerad** [comment]
@bobbinth I have made the suggested changes.

LMK if you prefer having separate functions as opposed to the `Option<DecoratorList>` for `ensure_block()`.

File: assembly/src/assembler/mast_forest_builder.rs [hunk 1]
```diff
@@ -58,6 +59,63 @@ impl MastForestBuilder {
...
+
+    /// Adds a syscall node to the forest, and returns the [`MastNodeId`] associated with it.
+    pub fn ensure_syscall(&mut self, callee: MastNodeId) -> Result<MastNodeId, MastForestError> {
+        self.ensure_node(MastNode::new_syscall(callee, &self.mast_forest))
+    }
+
+    /// Adds a dynexec node to the forest, and returns the [`MastNodeId`] associated with it.
+    pub fn ensure_dynexec(&mut self) -> Result<MastNodeId, MastForestError> {
+        self.ensure_node(MastNode::new_dynexec())
+    }
```
Comment (bobbinth) [inline assembly/src/assembler/mast_forest_builder.rs:112]: I would name this `ensure_dyn()` and, actually, I think `MastNode::new_dynexec()` should be `MastNode::new_dyn()` as well.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one small comment inline. After it is addressed, we can merge.

> LMK if you prefer having separate functions as opposed to the `Option<DecoratorList>` for `ensure_block()`.

I think the way you have it now looks good - so, no need to change anything in this regard.

File: assembly/src/assembler/mast_forest_builder.rs (hunk 1 above)
Comment (sergerad) [reply assembly/src/assembler/mast_forest_builder.rs]: Have updated both function names now.

**Comment by sergerad** [comment]
Just wondering whether we want to do the same for `MastForest::add_node()` which is used similarly. E.G:
```rust
let dyn_node_id = {
        let node = MastNode::new_dyn();
        mast_forest.add_node(node).unwrap()
    }
```

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

> Just wondering whether we want to do the same for `MastForest::add_node()`

Yes - that would be great! But I'd do it in a separate PR.

**Pull request merged**